### PR TITLE
Fix infinite loop in `process_inbox` in case of failed messages

### PR DESCRIPTION
**Pull request by ma2bd**
## Motivation

#1410 introduced some logic to skip messages that fails during block construction. However, failing messages that were not picked often stayed "pending" in the local node, causing `process_inbox` to loop.

## Proposal

Introduce "message actions", that is, the ability to explicitly reject a message. This was planned anyway as part of #990.

## Test Plan

CI (see modified unit test)

## Release Plan

- The system API are changed: Need to bump the major/minor version number in the next release of the crates.

### CODE REVIEW

File: linera-chain/src/chain.rs [hunk 1]
```diff
@@ -393,7 +393,7 @@ where
             let OutgoingMessage {
                 destination,
                 authenticated_signer,
-                is_skippable,
+                is_protected,
```
Comment (afck) [inline linera-chain/src/chain.rs:396]: Do we need that field, if it depends only on the message variant?

File: linera-chain/src/lib.rs [hunk 2]
```diff
@@ -71,15 +71,24 @@ pub enum ChainError {
...
+    CannotRejectMessage {
         chain_id: ChainId,
         origin: Box<Origin>,
         event: Event,
     },
     #[error(
-        "Incoming message in block proposed to {chain_id:?} has timestamp {message_timestamp:},
-        which is later than the block timestamp {block_timestamp:}."
+        "Block proposed to {chain_id:?} is attempting to skip a message \
+             that cannot be skipped: {event:?}"
```
Comment (afck) [inline linera-chain/src/lib.rs:82]:
```suggestion
         that cannot be skipped: {event:?}"
```

File: linera-core/src/client.rs [hunk 3]
```diff
@@ -1189,16 +1189,28 @@ where
...
+                        error!("Protected incoming message failed to execute locally: {message:?}");
+                    } else {
+                        // Reject the faulty message from the block and continue.
+                        // TODO(#1420): This is potentially a bit heavy-handed for
+                        // retryable errors.
+                        info!(
+                            "Message received from {:?} failed to execute locally \
+                               and will be rejected: {error}",
+                            message.origin
+                        );
```
Comment (afck) [inline linera-core/src/client.rs:1210]:
```suggestion
                        info!(
                            "Message received from {:?} failed to execute locally \
                             and will be rejected: {error}",
                            message.origin
                        );
```

Or:

```suggestion
                        info!(
                            %error, origin = ?message.origin,
                            "Message failed to execute locally and will be rejected."
                        );
```

**Review by afck — APPROVED** [review]

File: linera-chain/src/chain.rs (hunk 1 above)
Comment (ma2bd) [reply linera-chain/src/chain.rs:396]: Only system messages that never fails can be "protected" (counter-example: `Withdraw` which now correctly fails after ).

The notion of "protected" messages grew on me after I realized that:
* It helps with error handling during block construction (until we have something better).
* Together with fallback leaders, this will ensure that messages are eventually executed (within a reasonable timeframe). ("mandatory messages" require messages to be accepted in the next block but the implementation is more complex. We may not even need them in the end.)

File: linera-chain/src/lib.rs (hunk 2 above)
Comment (ma2bd) [reply linera-chain/src/lib.rs:82]: I believe `rustfmt` did that. Let me try again.

File: linera-chain/src/chain.rs (hunk 1 above)
Comment (afck) [reply linera-chain/src/chain.rs:396]: (Yes, I'm just saying it might not have to be a field in `OutgoingMessage`; it could just be a function, computed from the message variant.)

Comment (ma2bd) [reply linera-chain/src/chain.rs:396]: But we want the knowledge about system commands to be localized un system.rs right?

Comment (afck) [reply linera-chain/src/chain.rs:396]: Makes sense.
I guess `SystemMessage` could have an `is_protected` method (so the user doesn't need to know about the variants); but I'm fine with the field, too, of course!
